binlog_streamer.go

```diff
@@ -174,109 +266,47 @@ func (s *BinlogStreamer) Run() {
+		es.isEventPositionValid = true
 
-			s.logger.WithFields(logrus.Fields{
-				"new_position":  evPosition.Pos,
-				"new_filename":  evPosition.Name,
-				"last_position": s.lastStreamedBinlogPosition.Pos,
-				"last_filename": s.lastStreamedBinlogPosition.Name,
-			}).Info("binlog file rotated")
-		case *replication.FormatDescriptionEvent:
-			// This event is sent:
-			//   1) when our replication client connects to mysql
-			//   2) at the beginning of each binlog file
-			//
-			// For (1), if we are starting the binlog from a position that's greater
-			// than BIN_LOG_HEADER_SIZE (currently, 4th byte), this event's position
-			// is explicitly set to 0 and should not be considered valid according to
-			// the mysql source. See:
-			// https://github.com/percona/percona-server/blob/93165de1451548ff11dd32c3d3e5df0ff28cfcfa/sql/rpl_binlog_sender.cc#L1020-L1026
-			isEventPositionValid = ev.Header.LogPos != 0
-		case *replication.RowsQueryEvent:
-			// A RowsQueryEvent will always precede the corresponding RowsEvent
-			// if binlog_rows_query_log_events is enabled, and is used to get
-			// the full query that was executed on the master (with annotations)
-			// that is otherwise not possible to reconstruct
-			query = ev.Event.(*replication.RowsQueryEvent).Query
-		case *replication.RowsEvent:
-			err = s.handleRowsEvent(ev, query)
+		eventTypeString := ev.Header.EventType.String()
+		if handler, ok := s.eventHandlers[eventTypeString]; ok {
+			query, err = handler(ev, query, &es)
 			if err != nil {
-				s.logger.WithError(err).Error("failed to handle rows event")
+				s.logger.WithError(err).Error("failed to handle event")
 				s.ErrorHandler.Fatal("binlog_streamer", err)
 			}
-		case *replication.XIDEvent, *replication.GTIDEvent:
-			// With regards to DMLs, we see (at least) the following sequence
-			// of events in the binlog stream:
-			//
-			// - GTIDEvent  <- START of transaction
-			// - QueryEvent
-			// - RowsQueryEvent
-			// - TableMapEvent
-			// - RowsEvent
-			// - RowsEvent
-			// - XIDEvent   <- END of transaction
-			//
-			// *NOTE*
-			//
-			// First, RowsQueryEvent is only available with `binlog_rows_query_log_events`
-			// set to "ON".
-			//
-			// Second, there will be at least one (but potentially more) RowsEvents
-			// depending on the number of rows updated in the transaction.
-			//
-			// Lastly, GTIDEvents will only be available if they are enabled.
-			//
-			// As a result, the following case will set the last resumable position for
-			// interruption to EITHER the start (if using GTIDs) or the end of the
-			// last transaction
-			isEventPositionResumable = true
-
-			// Here we also reset the query event as we are either at the beginning
-			// or the end of the current/next transaction. As such, the query will be
-			// reset following the next RowsQueryEvent before the corresponding RowsEvent(s)
-			query = nil
+		} else {
+			query, err = s.defaultEventHandler(ev, query, &es)
```

Comment:
So the `defaultEventHandler` fires events to the `eventListeners`, but if one were to override the `eventHandler`, `eventListeners` will no longer be called? This also means that the default event listener that handles the row event (by sending it to the BinlogWriter) will no longer be called. You would have to manually reregister `handleRowEvent` as a eventHandler? This feels very confusing and somewhat error prone? Are there any cases where we don't want to row events to be handled as well?